learntools/ml-insights/nbs/tut4.ipynb

```diff
@@ -0,0 +1,259 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**[Machine Learning for Insights Home Page](kaggle.com/learn/ml-insights).**\n",
+    "\n",
+    "---\n",
+    "\n",
+    "# Shap Values\n",
+    "\n",
+    "You've seen (and tried) techniques to extract the general insights that a model learns. But these techniques don't help if you want to understand how a model made any single prediction. Our models can capture very complex relationships and interactions, so the average effect shown in a partial dependence plot may not resemble the effect of a feature for a single prediction we care about.\n",
+    "\n",
+    "Shap Values (also known as Shapley Additive Explanations) let you break down the impact of each feature for any single prediction. Where could you use this?\n",
+    "\n",
+    "- A model says a bank shouldn't loan someone money, and the bank is legally required to explain the basis for the loan rejection\n",
+    "- A healthcare providers wants to identify what factors are driving each patient's risk of some disease so they can directly address those risk factors\n",
+    "\n",
+    "You'll learn how Shap Values help explain individual predictions in this lesson. In the next lesson, you'll see how these can be aggregated into powerful model-level insights.\n",
+    "\n",
+    "# How They Work\n",
+    "\n",
+    "\n",
+    "It is easy to explain individual predictions when using very simple models.  For example, consider the equation\n",
+    "\n",
+    "```prediction = 10 + X1 + 2*x2```\n",
+    "\n",
+    "and consider a case where the **average** values of x1 and x2 are 0. Making predictions for when x1 and x2 are both 3, we could reasonably claim\n",
+    "- Our prediction increased by 3 because x1 is 3 (instead of 0)\n",
+    "- Our prediction increased by 6 because x2 is 3 (instead of 0)\n",
+    "\n",
+    "Summing these together, we'd claim `our prediction is 9 greater than average because x1 and x2 are both 3.` This is correct, as the prediction for average values is 10, and the prediction for these values is 19.\n",
+    "\n",
+    "But most machine learning models are more complex due to interactions between variables. We can illustrate the difficulty caused by interactions with another simple algebraic equation:\n",
+    "\n",
+    "```prediction = 10 + x1 + 2*x2 + x1*x2```\n",
+    "\n",
+    "What is the impact of having `x2` be 3 instead of it's average value of 0?  \n",
+    "\n",
+    "If we evaluate this when `x1` is 0, then the effect of x2 is 6.  But if we consider it when `x1` is 3, we must consider the last term as well, and the total impact of `x2` is 15.\n",
```

Comment:
Reading up to around this point, I'm starting to get confused. The original pitch was that we're going to learn a technique that will help us understand the reasons for an individual prediction.

Sure, the impact of having `x2` be 3 may change depending on whether `x` is 0, or 3, or whatever. But if we're reasoning about one specific instance, then we we have one specific value of `x1` to work with, no?